common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -177,478 +174,120 @@ public final boolean recycle(T o, Handle<T> handle) {
+        void push(DefaultHandle<T> node) {
+            if (node.next != null) {
                 throw new IllegalStateException("recycled already");
             }
-            item.recycleId = item.lastRecycledId = OWN_THREAD_ID;
-
-            int size = this.size;
-            if (size >= maxCapacity || dropHandle(item)) {
-                // Hit the maximum capacity or should drop - drop the possibly youngest object.
-                return;
-            }
-            if (size == elements.length) {
-                elements = Arrays.copyOf(elements, min(size << 1, maxCapacity));
-            }
-
-            elements[size] = item;
-            this.size = size + 1;
-        }
-
-        private void pushLater(DefaultHandle<?> item, Thread thread) {
-            // we don't want to have a ref to the queue as the value in our weak map
-            // so we null it out; to ensure there are no races with restoring it later
-            // we impose a memory ordering here (no-op on x86)
-            Map<Stack<?>, WeakOrderQueue> delayedRecycled = DELAYED_RECYCLED.get();
-            WeakOrderQueue queue = delayedRecycled.get(this);
-            if (queue == null) {
-                if (delayedRecycled.size() >= maxDelayedQueues) {
-                    // Add a dummy queue so we know we should drop the object
-                    delayedRecycled.put(this, WeakOrderQueue.DUMMY);
-                    return;
-                }
-                // Check if we already reached the maximum number of delayed queues and if we can allocate at all.
-                if ((queue = WeakOrderQueue.allocate(this, thread)) == null) {
-                    // drop object
-                    return;
-                }
-                delayedRecycled.put(this, queue);
-            } else if (queue == WeakOrderQueue.DUMMY) {
-                // drop object
-                return;
+            int i = 0;
+            for (; i < genMaxId.length && genMaxId[i] < node.genId; i++) {
+                // find generation index
             }
-
-            queue.add(item);
-        }
-
-        boolean dropHandle(DefaultHandle<?> handle) {
-            if (!handle.hasBeenRecycled) {
-                if ((++handleRecycleCount & ratioMask) != 0) {
-                    // Drop the object.
-                    return true;
-                }
-                handle.hasBeenRecycled = true;
+            node.next = genHead[i];
+            genHead[i] = node.selfRef;
+        }
+
+        T popOrCreate() {
+            DefaultHandle<T> handle = pop();
+            if (handle != null) {
+                return handle.value;
+            } else if ((allocationCount++ & 63) == 0) {
+                // Create recyclable objects at a low rate. If they are helpful, they will accumulate
+                handle = new DefaultHandle<T>(parent);
+                handle.value = parent.newObject(handle);
+                return handle.value;
+            } else {
```

Comment:
you can remove the else.

Reply:
Done. While I generally prefer avoiding the unnecessary else, I'm not sure it is as clear in this case.